cnd/src/seed.rs

```diff
@@ -12,82 +12,109 @@ use std::{
 };
 use thiserror;
 
-pub const SEED_LENGTH: usize = 32;
+const SEED_LENGTH: usize = 32;
+pub type Seed = [u8; SEED_LENGTH];
+
+/// We create a `RootSeed` either randomly or by reading in the PEM file from
+/// disk.  This `RootSeed` is used to generate a per swap `SwapSeed` which is
+/// then use as the secret source for deriving redeem/refund identities.
+/// `RootSeed` and `SwapSeed` are the same underlying type (`Seed`), they exist
+/// solely to allow the compiler to provide us with type safety.
+
+pub trait DeriveSwapSeed {
+    fn derive_swap_seed(&self, id: SwapId) -> SwapSeed;
+}
+
+impl DeriveSwapSeed for RootSeed {
+    fn derive_swap_seed(&self, id: SwapId) -> SwapSeed {
+        SwapSeed(self.sha256_with_seed(&[b"SWAP", id.0.as_bytes()]))
+    }
+}
+
 #[derive(Clone, Copy, Serialize, Deserialize, PartialEq)]
-pub struct Seed(#[serde(with = "hex_serde")] [u8; SEED_LENGTH]);
+pub struct RootSeed(#[serde(with = "hex_serde")] Seed);
+
+impl fmt::Debug for RootSeed {
```

Comment:
If you make `Seed` an actual newtype, you could implement `Debug` and `Display` once for that type and you would not need to repeat it here.

I would then forward the `Debug` and `Display` impls to the inner one.

Reply:
Awesome, done as suggested.